Feat: 주문완료 직후와 전체내역 메일 발송 / 주문 완료시 주문 일괄 처리

ky1nonly opened the pull request:
## 🛰️ Issue Number
#7 

## 🪐 작업 내용

* OrderServiceTests는 임시로 삭제하고 올렸습니다!

## 주문완료 직후 메일 전송 
- OrderService :  주문 직후 이벤트 생성
- OrderMailEventListener : 주문완료 메일 전송 메서드 호출
```
public void onOrderCreated(OrderCreatedEvent event) {

        try {
            mailService.sendOrderConfirmation(event.order());
        } catch (MessagingException e) {
            log.error("주문 확인 메일 발송 실패 – orderId={}, 에러={}",
                    event.order().getId(), e.getMessage(), e);
        }
    }

```

## 오후2시 전체주문내역 메일 전송
- DailyOrderSummarySchedulerService : 스케쥴러가 전체주문내역 메일 전송 메서드 호출
- 전체주문내역을 Order 엔티티 그대로 넘기지 않고  OrderSummaryDto로 전달하도록 수정
```
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class OrderDto {

    private String email;
    private String address;
    private String postNum;

    @Builder
    public OrderDto(String email, String address, String postNum) {
        this.email = email;
        this.address = address;
        this.postNum = postNum;
    }
}
```

### 전체주문내역 메일 전송 완료 후 
- orderRepository에서   isProcessed=false로 일괄 업데이트
```
        List<Long> ids = orders.stream()
                .map(Order::getId)
                .toList();

        int updated = orderRepository.markProcessedFalseByIdIn(ids);
        log.info("isProcessed=false 업데이트 {}건 완료", updated);
```

## 오후 2시 일괄메일 전송 테스트 통과
## 📚 Reference


## ✅ Check List
- [x] 코드가 정상적으로 컴파일되나요?
- [x] 테스트 코드를 통과했나요?
- [x] merge할 브랜치의 위치를 확인했나요?
- [x] Label을 지정했나요?

ky1nonly committed:
feat: 이메일 전송 기능 구현(#7)

ky1nonly committed:
Reafactort: OrderSummaryDto로 전체주문내역 전달 (#17)

peng255 commented on src/main/java/io/sleepyhoon/project1/service/DailyOrderSummarySchedulerService.java:
```diff
@@ -0,0 +1,70 @@
+
+
+        List<Order> orders =
+                orderRepository.findByIsProcessedTrueAndOrderedAtBetween(start, end);
```
P1: 꼭 반영해주세요 (Request changes)
isProcessed가 False인 메일로 수정해주셔야 잘 작동될 것 같습니다 😄

ky1nonly replied:
반영하였습니다! isProcessed의 의미를 헷갈려 실수했네요ㅎㅎ 잡아주셔서 감사합니다!

peng255 reviewed (commented):
처음 배우는 내용으로 구현하시느라 너무 고생 많으셨습니다!! 🥲

ky1nonly committed:
fix:isProcessed T/F 다 반대로 썼던 부분 수정

sleepyhoon commented on src/main/java/io/sleepyhoon/project1/service/DailyOrderSummarySchedulerService.java:
```diff
@@ -0,0 +1,70 @@
+
+        ZoneId zone = ZoneId.of("Asia/Seoul");
+        LocalDate today = LocalDate.now(zone);
+        LocalDateTime start = today.atTime(2, 0);
+        LocalDateTime end   = today.atTime(14, 0);
```
### 꼭 반영해주세요 (Request changes)
현재 요구사항은 어제 14시부터 오늘 14시에 해당하는 주문을 처리해야하는데, 지금은 같은 날 새벽 2시부터 14시까지 해당하는 주문을 조회하고 있는 듯 합니다... 수정해주세요!

ky1nonly replied:
앞서 미리 말씀해주셨던  부분인데 반영을 놓쳤네요! 반영했습니다

sleepyhoon commented on src/main/java/io/sleepyhoon/project1/service/DailyOrderSummarySchedulerService.java:
```diff
@@ -0,0 +1,70 @@
+                orderRepository.findByIsProcessedFalseAndOrderedAtBetween(start, end);
+
+        if (orders.isEmpty()) {
+            log.info("오늘 02-14시 주문 없음 → 메일 발송 생략");
```
여기도 시간 수정해주세요!

ky1nonly replied:
오 로그까지 고쳐야할 부분을 직접 찾아주시다니 정말 감사합니다. 확인했습니다!

sleepyhoon commented on src/main/java/io/sleepyhoon/project1/service/DailyOrderSummarySchedulerService.java:
```diff
@@ -0,0 +1,70 @@
+    private final OrderMailService orderMailService;
+
+
+    @Scheduled(cron = "0 0 14 * * *", zone = "Asia/Seoul")
```
스케줄링 구현해주셨네요! 감사합니다!

sleepyhoon reviewed (commented):
메일 전송 로직 구현하기 쉽지 않았을텐데 너무 고생하셨습니다!

ky1nonly committed:
Fix:주문건 가져오는 시간 14시로 변경

ky1nonly committed:
Merge branch 'dev' into feat/7-send-emails

chw0912 commented on src/main/java/io/sleepyhoon/project1/dao/OrderRepository.java:
```diff
@@ -30,4 +33,13 @@ public interface OrderRepository extends JpaRepository<Order, Long> {
+
+    List<Order> findByIsProcessedFalseAndOrderedAtBetween(
+            LocalDateTime start, LocalDateTime end);
+
+    @Modifying
+    @Transactional
+    @Query("UPDATE Order o SET o.isProcessed = true WHERE o.id IN :ids")
+    int markProcessedTrueByIdIn(@Param("ids") List<Long> ids);
 }
```
Service 계층에서 `@Transactional`을 붙여주는 걸로 알고 있는데 (저도 잘 몰라서..ㅎ)
Repository계층에 사용하신 이유가 있으실까요?

ky1nonly replied:
말씀대로 지금 상황에서는 Repository계층에 트랜잭션이 불필요한 것 같습니다! 
 이부분을 만들던 중에 나중에 변경할 사항을 생각하며 적어놓고 제거하는걸 깜빡한 것 같아요. 

생각하던 추가 기능은 일괄메일 전송 중도실패시 전체 롤백이 아닌 메일 전송에 성공한 주문과 실패한 주문을 따로 처리하는 것이었는데, 이부분을 구현하지 못한다면 Repository 의 @ Transactional은 빼는게 좋을 것 같네요! 
짚어주셔서 감사합니다 ㅎㅎ

chw0912 commented on src/main/java/io/sleepyhoon/project1/service/OrderMailService.java:
```diff
@@ -0,0 +1,60 @@
+    private final MailTemplateService templateService;
+
+
+    public void sendOrderConfirmation(Order order) throws MessagingException {
+        MimeMessage msg = mailSender.createMimeMessage();
+        MimeMessageHelper helper = new MimeMessageHelper(
+                msg, true, StandardCharsets.UTF_8.name());
+
```
이 부분은 아래 `sendDailyOrderSummary`에서도 반복되는 구조인 것 같습니다.
나중에 리팩토링 할 때 메서드로 구현해봐도 좋을 것 같습니다! 👍

ky1nonly replied:
오 좋은 의견 감사합니다 반영해보겠습니다🙌

chw0912 reviewed (commented):
스케쥴링 구현하는 방법을 배우고 갑니다! 많이 늦었지만 리뷰 남겼습니다..ㅎ

ky1nonly committed:
Merge branch 'dev' of https://github.com/prgrms-be-devcourse/NBE5-7-1-Team10 into feat/7-send-emails

ky1nonly committed:
A
Merge remote-tracking branch 'origin/feat/7-send-emails' into feat/7-send-emails

ky1nonly committed:
fix:괄호 오타 수정

ky1nonly committed:
test: 일일 일괄메일전송 테스트 성공
